src/napari/_qt/_tests/test_qt_viewer.py

```diff
@@ -719,6 +732,206 @@ def test_leaks_labels(qtbot, make_napari_viewer):
+    view = viewer.window._qt_viewer
+
+    # Track various objects using weakrefs
+    layer_refs = []
+    data_refs = []
+    visual_refs = []
+    widget_refs = []
+
+    # Test different layer types
+    layer_types = [
+        ('image', np.random.rand(10, 10)),
+        ('labels', (np.random.rand(10, 10) * 10).astype(np.uint8)),
+        ('points', np.random.rand(10, 2)),
+        ('shapes', [np.array([[0, 0], [10, 0], [10, 10], [0, 10]])]),
+    ]
+
+    # Add layers and collect references
+    for layer_type, data in layer_types:
+        if layer_type == 'image':
+            layer = viewer.add_image(data)
+        elif layer_type == 'labels':
+            layer = viewer.add_labels(data)
+        elif layer_type == 'points':
+            layer = viewer.add_points(data)
+        elif layer_type == 'shapes':
+            layer = viewer.add_shapes(data)
+
+        # Store weak references
+        layer_refs.append(weakref.ref(layer))
+
+        # For shapes, layer.data is a list which cannot be weakly referenced
+        if layer_type != 'shapes':
+            data_refs.append(weakref.ref(layer.data))
+
+        # Store visual references if available
+        if (
+            hasattr(view.canvas, 'layer_to_visual')
+            and layer in view.canvas.layer_to_visual
+        ):
+            visual_refs.append(weakref.ref(view.canvas.layer_to_visual[layer]))
+
+        # Store widget references if available
+        if (
+            hasattr(view.controls, 'widgets')
+            and layer in view.controls.widgets
+        ):
+            widget_refs.append(weakref.ref(view.controls.widgets[layer]))
+
+    # Ensure all layers are properly initialized
+    qtbot.wait(50)
+    QApplication.processEvents()
+
+    # Remove all layers
+    viewer.layers.clear()
+
+    # Ensure Qt events are processed
+    qtbot.wait(100)
+    QApplication.processEvents()
+    qtbot.wait(100)
+
+    # Close the viewer to ensure all Qt resources are released
+    viewer.close()
+    qtbot.wait(50)
+
+    # Delete viewer reference
+    del viewer
+    del view
+    qtbot.wait(50)
+
+    # Force garbage collection multiple times
+    for _ in range(len(layer_refs)):
+        gc.collect()
+        qtbot.wait(10)
+
+    # Check that all references are cleaned up
+    for i, visual_ref in enumerate(visual_refs):
+        assert visual_ref() is None, f'Visual {i} was not garbage collected'
+
+    for i, widget_ref in enumerate(widget_refs):
+        assert widget_ref() is None, f'Widget {i} was not garbage collected'
+
+    for i, layer_ref in enumerate(layer_refs):
+        # TODO: Shapes layers have a known memory leak issue where they don't get
+        # properly garbage collected. This is likely due to circular references
+        # in the Shapes layer implementation (event handlers, callbacks, or VisPy
+        # visual components). Skip this assertion for now until the underlying
+        # issue is fixed. See: https://github.com/napari/napari/issues/XXXX
```

Comment:
Issue doesn't exist yet or is this a todo?

Reply:
This was pretty much a TODO to myself while mucking around with this.